README.md
# Replicator

[![Build Status](https://travis-ci.org/elsevier-core-engineering/replicator.svg?branch=master)](https://travis-ci.org/elsevier-core-engineering/replicator) [![Go Report Card](https://goreportcard.com/badge/github.com/elsevier-core-engineering/replicator)](https://goreportcard.com/report/github.com/elsevier-core-engineering/replicator) [![Join the chat at https://gitter.im/els-core-engineering/replicator/Lobby](https://badges.gitter.im/els-core-engineering/replicator/Lobby.svg)](https://gitter.im/els-core-engineering/replicator?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![GoDoc](https://godoc.org/github.com/elsevier-core-engineering/replicator?status.svg)](https://godoc.org/github.com/elsevier-core-engineering/replicator)

Replicator is a fast and highly concurrent Go daemon that provides dynamic scaling of [Nomad](https://github.com/hashicorp/nomad) jobs and worker nodes.

- Replicator job scaling policies are configured as [meta parameters](https://www.nomadproject.io/docs/job-specification/meta.html) within the job specification. A job scaling policy allows scaling constraints to be defined per task-group. Currently supported scaling metrics are CPU and Memory; there are plans for additional metrics as well as different metric backends in the future. Details of configuring job scaling and other important information can be found on the Replicator [Job Scaling wiki page](https://github.com/elsevier-core-engineering/replicator/wiki/Job-Scaling).

- Replicator supports dynamic scaling of multiple, distinct cluster worker nodes in an AWS autoscaling group. Worker pool autoscaling is configured through Nomad client [meta parameters](https://www.nomadproject.io/docs/agent/configuration/client.html#meta). Details of configuring worker pool scaling and other important information can be found on the Replicator [Cluster Scaling wiki page](https://github.com/elsevier-core-engineering/replicator/wiki/Cluster-Scaling).

*At present, worker pool autoscaling is only supported on AWS, however, future support for GCE and Azure are planned using the Go factory/provider pattern.*

### Download

Pre-compiled releases for a number of platforms are available on the [GitHub release page](https://github.com/elsevier-core-engineering/replicator/releases). Docker images are also available from the elsce [Docker Hub page](https://hub.docker.com/r/elsce/replicator/).

## Running

Replicator can be run in a number of ways; the recommended way is as a Nomad service job either using the [Docker driver](https://www.nomadproject.io/docs/drivers/docker.html) or the [exec driver](https://www.nomadproject.io/docs/drivers/exec.html). There are example Nomad [job specification files](https://github.com/elsevier-core-engineering/replicator/tree/master/example-jobs) available as a starting point.

It's recommended to take a look at the agent [configuration options](https://github.com/elsevier-core-engineering/replicator/wiki/Agent-Command) to configure Replicator to run best in your environment.

Replicator is fully capable of running as a distributed service; using [Consul sessions](https://www.consul.io/docs/internals/sessions.html) to provide leadership locking and exclusion. State is also written by Replicator to the Consul KV store, allowing Replicator failures to be handled quickly and efficiently.

An example Nomad client configuration that can be used to enable autoscaling on the worker pool:
```hcl
bind_addr = "0.0.0.0"
client {
  enabled =  true
  meta {
    "replicator_cooldown"            = 400
    "replicator_enabled"              = true
    "replicator_node_fault_tolerance" = 1
    "replicator_notification_uid"     = "REP2"
    "replicator_provider"             = "aws"
    "replicator_region"               = "us-east-1"
    "replicator_retry_threshold"      = 3
    "replicator_scaling_threshold"    = 3
    "replicator_worker_pool"          = "container-node-public-prod"
  }
}
```

An example job which has autoscaling enabled:
```hcl
job "example" {
  datacenters = ["dc1"]
  type        = "service"

  update {
    max_parallel = 1
    stagger      = "10s"
  }

  group "cache" {
    count = 3

    meta {
      "replicator_max"               = 10
      "replicator_cooldown"          = 50
      "replicator_enabled"           = true
      "replicator_min"               = 1
      "replicator_retry_threshold"   = 1
      "replicator_scalein_mem"       = 30
      "replicator_scalein_cpu"       = 30
      "replicator_scaleout_mem"      = 80
      "replicator_scaleout_cpu"      = 80
      "replicator_notification_uid"  = "REP1"
    }

    task "redis" {
      driver = "docker"
      config {
        image = "redis:3.2"
        port_map {
          db = 6379
        }
      }

      resources {
        cpu    = 500 # 500 MHz
        memory = 256 # 256MB
        network {
          mbits = 10
          port "db" {}
        }
      }

      service {
        name = "global-redis-check"
        tags = ["global", "cache"]
        port = "db"
        check {
          name     = "alive"
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
    }
  }
}
```

### Permissions

Replicator requires permissions to Consul and the AWS (the only currently supported cloud provider) API in order to function correctly. The Consul ACL token is passed as a configuration parameter and AWS API access should be granted using an EC2 instance IAM role. Vault support is planned for the near future, which will change the way in which permissions are managed and provide a much more secure method of delivering these.

#### Consul ACL Token Permissions

If the Consul cluster being used is running ACLs; the following ACL policy will allow Replicator the required access to perform all functions based on its default configuration:

```hcl
key "" {
  policy = "read"
}
key "replicator/config" {
  policy = "write"
}
node "" {
  policy = "read"
}
node "" {
  policy = "write"
}
session "" {
  policy = "read"
}
session "" {
  policy = "write"
}
```

#### AWS IAM Permissions

Until Vault integration is added, the instance pool which is capable of running the Replicator daemon requires the following IAM permissions in order to perform worker pool scaling:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "AuthorizeAutoScalingActions",
            "Action": [
                "autoscaling:DescribeAutoScalingGroups",
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeScalingActivities",
                "autoscaling:DetachInstances",
                "autoscaling:UpdateAutoScalingGroup"
            ],
            "Effect": "Allow",
            "Resource": "*"
        },
        {
            "Sid": "AuthorizeEC2Actions",
            "Action": [
                "ec2:DescribeInstances",
                "ec2:DescribeRegions",
                "ec2:TerminateInstances",
                "ec2:DescribeInstanceStatus"
            ],
            "Effect": "Allow",
            "Resource": "*"
        }
    ]
}
```

### Commands

Replicator supports a number of commands (CLI) which allow for the easy control and manipulation of the replicator binary. In-depth documentation about each command can be found on the Replicator [commands wiki page](https://github.com/elsevier-core-engineering/replicator/wiki/Commands).

#### Command: `agent`

The `agent` command is the main entry point into Replicator. A subset of the available replicator agent configuration can optionally be passed in via CLI arguments and the configuration parameters passed via CLI flags will always take precedent over parameters specified in configuration files.

Detailed information regarding the available CLI flags can be found in the Replicator [agent command wiki page](https://github.com/elsevier-core-engineering/replicator/wiki/Agent-Command).

#### Command: `failsafe`

The `failsafe` command is used to toggle failsafe mode across the pool of Replicator agents. Failsafe mode prevents any Replicator agent from taking any scaling actions on the resource placed into failsafe mode.

Detailed information about failsafe mode operations and the available CLI options can be found in the Replicator [failsafe command wiki page](https://github.com/elsevier-core-engineering/replicator/wiki/Failsafe-Command).

#### Command: `init`

The `init` command creates example job scaling and worker pool scaling meta documents in the current directory. These files provide a starting example for configuring both scaling functionalities.

#### Command: `version`

The `version` command displays build information about the running binary, including the release version.

## Frequently Asked Questions

### When does Replicator adjust the size of the worker pool?

Replicator will dynamically scale-in the worker pool when:
- Resource utilization falls below the capacity required to run all current jobs while sustaining the configured node fault-tolerance. When calculating required capacity, Replicator includes scaling overhead required to increase the count of all running jobs by one.
- Before removing a worker node, Replicator simulates capacity thresholds if we were to remove a node. If the new required capacity is within 10% of the current utilization, Replicator will decline to remove a node to prevent thrashing.

Replicator will dynamically scale-out the worker pool when:
- Resource utilization exceeds or closely approaches the capacity required to run all current jobs while sustaining the configured node fault-tolerance. When calculating required capacity, Replicator includes scaling overhead required to increase the count of all running jobs by one.

### When does Replicator perform scaling actions against running jobs?

Replicator will dynamically scale a job when:
- A valid scaling policy for the job task-group is present within the job specification meta parameters and has the enabled flag set to true.
- A job specification can consist of multiple groups, each group can contain multiple tasks. Resource allocations and count are specified at the group level.
- Replicator evaluates scaling thresholds against the resource requirements defined within a group task. If any task within a group is found to violate the scaling thresholds, the group count will be adjusted accordingly.

## Contributing

Contributions to Replicator are very welcome! Please refer to our [contribution guide](https://github.com/elsevier-core-engineering/replicator/blob/master/.github/CONTRIBUTING.md) for details about hacking on Replicator.

For questions, please check out the [Elsevier Core Engineering/replicator](https://gitter.im/els-core-engineering/replicator) room in Gitter.


CHANGELOG.md
## 1.1.0 (Unreleased)

NEW FEATURES:

* **API Framework**: Replicator now runs as an RPC server and implements an HTTP API framework. [GH-186]
* **Scaling Provider**: Replicator now implements a cloud provider model to support cluster scaling operations across multiple cloud providers. [GH-203]
* **OpsGenie Notifier**: Replicator now supports OpsGenie as a backend notifier. Thank you to @vladshub. [GH-225]
* **Worker Pool Configuration In Consul**: Replicator now supports the ability to store full worker pool configuration in the Consul Key/Value store allowing for minimal configuration to be stored in the node meta configuration parameters. [GH-204]
* **replicator_retry_threshold**: A new config flag added to allow configuration for retrying job scaling activites before entering failsafe. Thank you to @djenriquez [GH-261] 

BUG FIXES:

* Filter out Nomad jobs that do not include an `update` stanza to prevent segmentation violations when attempting to verify job scaling activities. Thank you to @burdandrei. [GH-209]
* Update init command to generate correct cluster_scaling meta tags. Thank you to @burdandrei. [GH-210]
* Pass job scaling verification and processing methods the job ID rather than name to handle jobs where the name differs from the ID. Thank you to @burdandrei. [GH-212]
* Prevent processing of Nomad system jobs. [GH-220]
* Prevent panic when Replicator fails to determine the IP address of a node during worker pool scale-in operations. Thank you to @burdandrei. [GH-222]
* Prevent errors when Replicator attempts a cluster scale-in operation while the prioritized scaling metric is `ScalingMetricDsik`. [GH-240]
* Increase timeout for ASG scaling result. [GH-258]
* Use `time.NewTicker` rather than `time.Tick` for garbage collection issues and timeouts. [GH-262]
* Continue on with node termination even if the ASG detach fails. [GH-269]
* Fix issue caused by changes Nomad 0.8 where deployments can now return null. [GH-270]


IMPROVEMENTS:

* Remove explicit min/max thresholds for cluster scaling in favor of dynamic threshold discovery in the cloud provider. [GH-211]
* Remove all internal constants used in favor of exported constants from the Nomad structs package. [GH-213]
* Job scaling API calls now allow stale reads to reduce load on the Nomad leader. [GH-224]
* Replicator now implements concurrent scaling limits on both job and cluster scaling operations. [GH-223, GH-221]
* Replicator now implements a `status/leader` API endpoint for retrieving Replicator leader information. [GH-242]
* Logging messages to include date, time and zone [GH-254]

## 1.0.3 (22 September 2017)

BUG FIXES:

* Reset accumulated job scaling values during each scaling evaluation. Thank you to @comebackoneyear and @lifesum. [GH-202]
* Average job utilization from all allocations, not just the last one processed. Thank you to @comebackoneyear and @lifesum. [GH-202]

## 1.0.2 (15 September 2017)

BUG FIXES:

* Improve and fix node protection mechanism. [GH-197]

## 1.0.1 (14 September 2017)

BUG FIXES:

* Reset failure count when exiting failsafe [GH-190]
* Retry when Nomad returns an empty deployment ID [GH-191]

IMPROVEMENTS:

* Improve accuracy and detail of cluster_scaling telemetry metrics. [GH-194]
* Improve accuracy and detail of job_scaling telemetry metrics. [GH-193]
* Improve logging output of cluster scaling operations when logging set to `INFO`. [GH-195]

## 1.0.0 (11 September 2017)

IMPROVEMENTS:

* Nomad JobScaling now uses meta parameters within the job file for configuration [GH-139]
* Replicator now tracks scaling deployment to confirm success [GH-146]
* Introduce Node Discovery With Dynamic Refresh [GH-148]
* Replicator now supports setting failsafe per job group [GH-152]
* Replicator now supports concurrent cluster scaling of multiple worker pools [GH-157]

## 0.0.2 (03 August 2017)

IMPROVEMENTS:

* Replicator now supports persistent storage of state tracking information in
Consul Key/Value stores. This allows the daemon to restart and gracefully
resume where it left off and supports graceful leadership changes. [GH-92]
* Scaling cooldown threshold is calculated and tracked internally rather than
externally referencing the worker pool autoscaling group. [GH-91]
* Failed worker nodes are detached after maximum retry interval is reached for
troubleshooting. [GH-76]
* Add support for event notifications via PagerDuty [GH-66]
* Replicator will send a notification if failsafe mode is enabled [GH-131]
* New nodes are verified to have successfully joined the worker pool after a
cluster scaling operation is initiated. Replicator will retry until a healthy
node is launched or the maximum retry threshold is reached. [GH-62]
* Allow Dry Run Operations Against Job Scaling Documents. [GH-53]
* Add the `scaling_interval config` parameter to replicator agent. [GH-49]
* Increase Default Cluster Scaling Cooldown Period to 600s. [GH-56]
* Addition of CLI flags for all configuration parameters. [GH-61]
* Add support for `dev` CLI flag [GH-63]
* AWS `TerminateInstance` function now verifies instance has successfully
transitioned to a terminated state. [GH-80]
* Make use of telemetry configuration by sending key metrics. [GH-85]
* Replicator now runs leadership locking using Consul sessions +  KV [GH-101]
* Introduce distributed failsafe mode and new failsafe CLI command [GH-105]
* `cluster-scaling-theshold` parameter is now used to determine scaling safety [GH-115]

BUG FIXES:

* The global job scaling enabled flag is now evaluated before allowing job
scaling operations. [GH-81]
* Prevent Divide By Zero Panic When Replicator Detects No Healthy Worker
Nodes. [GH-55]
* Prevent Cluster Scaling Operations That Would Violate ASG Constraints.
[GH-142]

## 0.0.1 (02 May 2017)

- Initial release.


command/init.go
package command

import (
	"fmt"
	"io/ioutil"
	"os"
	"strings"
)

const (
	defaultJobScalingName     = "job_scaling.hcl"
	defaultClusterScalingName = "cluster_scaling.hcl"
)

// InitCommand is the command implamentation for init.
type InitCommand struct {
	Meta
}

// Help provides the help information for the init command.
func (c *InitCommand) Help() string {
	helpText := `
Usage: replicator init [options]

  Creates example job and cluster scaling configurations.

  General Options:

    -job-scaling
      Write a file which contains example job scaling configuration. This
      can be used directly within the Nomad job specification file to enable
      scaling for the desired job group.

    -cluster-scaling
      Write a file which contains example cluster scaling configuration.
      This can be adapted to your configuration management to enable
      cluster scaling.
`
	return strings.TrimSpace(helpText)
}

// Synopsis is provides a brief summary of the init command.
func (c *InitCommand) Synopsis() string {
	return "Create example Replicator job and cluster scaling configurations"
}

// Run triggers the init command to write the example.json file out to the
// current directory.
func (c *InitCommand) Run(args []string) int {

	var jobScaling, clusterScaling bool

	// Initialize command flags.
	flags := c.Meta.FlagSet("init", FlagSetClient)
	flags.Usage = func() { c.UI.Error(c.Help()) }

	// General configuration flags.
	flags.BoolVar(&jobScaling, "job-scaling", false, "")
	flags.BoolVar(&clusterScaling, "cluster-scaling", false, "")

	// Parse the passed CLI flags.
	if err := flags.Parse(args); err != nil {
		return 1
	}

	if jobScaling || len(args) == 0 {
		err := c.writeFile(defaultJobScalingName, defaultJobScalingDocument)
		if err != nil {
			c.UI.Error(fmt.Sprintf("Unable to write file %s: %v", defaultJobScalingName, err))
			return 1
		}
	}

	if clusterScaling || len(args) == 0 {
		err := c.writeFile(defaultClusterScalingName, defaultClusterScalingDocument)
		if err != nil {
			c.UI.Error(fmt.Sprintf("Unable to write file %s: %v", defaultClusterScalingName, err))
			return 1
		}
	}
	return 0
}

func (c *InitCommand) writeFile(file, content string) (err error) {

	// Check if the file already exists.
	if _, err = os.Stat(file); err != nil && !os.IsNotExist(err) {
		return err
	}

	if !os.IsNotExist(err) {
		return fmt.Errorf("scaling document already exists")
	}

	// Write the example file to the relative local directory where Replicator
	// was invoked from.
	if err = ioutil.WriteFile(file, []byte(content), 0660); err != nil {
		return err
	}

	c.UI.Info(fmt.Sprintf("Example scaling configuration written to %s", file))
	return
}

var defaultJobScalingDocument = strings.TrimSpace(`
meta {
  "replicator_cooldown"         = 50
  "replicator_enabled"          = true
  "replicator_max"              = 10
  "replicator_min"              = 1
  "replicator_notification_uid" = "REP1"
  "replicator_scalein_mem"      = 30
  "replicator_scalein_cpu"      = 30
  "replicator_scaleout_mem"     = 80
  "replicator_scaleout_cpu"     = 80
}
`)

var defaultClusterScalingDocument = strings.TrimSpace(`
meta {
  "replicator_cool_down"            = 400
  "replicator_enabled"              = true
  "replicator_node_fault_tolerance" = 1
  "replicator_notification_uid"     = "REP2"
  "replicator_provider"             = "aws"
  "replicator_region"               = "us-east-1"
  "replicator_retry_threshold"      = 3
  "replicator_scaling_threshold"    = 3
  "replicator_worker_pool"          = "container-node-public-prod"
}
`)
